File: src/main/java/co/herod/adbwrapper/AdbDeviceManager.java
package co.herod.adbwrapper;

import co.herod.adbwrapper.model.AdbDevice;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.annotations.CheckReturnValue;

import java.util.List;
import java.util.concurrent.TimeUnit;

@SuppressWarnings("WeakerAccess")
public class AdbDeviceManager {

    @CheckReturnValue
    public static AdbDevice getConnectedDevice() {
        return getConnectedDeviceSingle().blockingGet();
    }

    @CheckReturnValue
    public static Single<AdbDevice> getConnectedDeviceSingle() {
        return connectedDevices()
                .filter(AdbDevice::isConnectedDevice)
                .firstOrError()
                .retryWhen(handler -> handler.delay(1, TimeUnit.SECONDS))
                .timeout(10, TimeUnit.SECONDS);
    }

    @CheckReturnValue
    public static List<AdbDevice> getAllDevices() {
        return connectedDevices().toList().blockingGet();
    }

    private static Observable<AdbDevice> connectedDevices() {
        return Adb.devices();
    }
}

File: adb-kotlin/src/main/java/co/herod/adbwrapper/AdbDeviceManager.kt
/*
 * Copyright (c) 2018. Herod
 */

@file:Suppress("unused")

package co.herod.adbwrapper

import co.herod.adbwrapper.S.Companion.DEVICES
import co.herod.adbwrapper.model.AdbDevice
import co.herod.adbwrapper.processes.parseAdbString
import io.reactivex.Observable
import io.reactivex.annotations.CheckReturnValue
import java.util.concurrent.TimeUnit

object AdbDeviceManager {

    @JvmStatic
    @CheckReturnValue
    fun getDevice(): AdbDevice? = allDevices()
            .firstOrError()
            .retryWhen { it.delay(1, TimeUnit.SECONDS) }
            .timeout(10, TimeUnit.SECONDS)
            .onErrorReturn { null }
            .blockingGet()

    @JvmStatic
    @CheckReturnValue
    fun getConnectedDevice(): AdbDevice? = allDevices()
            .filter(AdbDevice::physical)
            .firstOrError()
            .retryWhen { it.delay(1, TimeUnit.SECONDS) }
            .timeout(10, TimeUnit.SECONDS)
            .blockingGet()

    @JvmStatic
    @CheckReturnValue
    fun getAllDevices(): List<AdbDevice> = allDevices()
            .toList()
            .blockingGet()
}

private fun allDevices(): Observable<AdbDevice> =
        AdbCommand.Builder()
                .setCommand(DEVICES)
                .observable()
                .filter { "\t" in it }
                .map(::parseAdbString)

